src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java

```diff
@@ -176,11 +180,12 @@ public static SegmentRunner create(
   @Override
   public void run() {
     boolean ran = false;
-    if (takeLead()) {
+    RepairSegment segment = context.storage.getRepairSegment(repairRunner.getRepairRunId(), segmentId).get();
+    if (takeLead(segment)) {
```

Comment:
During the design of this work I thought we agreed that there should be the opportunistic check before the lock is attempted…?

For example, making the method like
```
  @Override
  public void run() {
    boolean ran = false;
    try {
      RepairSegment segment = context.storage.getRepairSegment(repairRunner.getRepairRunId(), segmentId).get();
      JmxProxy coordinator = clusterFacade.connect(cluster, potentialCoordinators);
      if (canRepair(segment, coordinator) && takeLead(segment)) {
        try {
          ran = runRepair(coordinator);
        } finally {
          releaseLead(segment);
        }
      }
    } catch (ReaperException e) {
      LOG.warn("Failed to connect to a coordinator node for checking segment {}", segmentId, e);
    }
    if (ran) {
      long delay = intensityBasedDelayMillis(intensity);
      try {
        Thread.sleep(delay);
      } catch (InterruptedException e) {
        LOG.warn("Slept shorter than intended delay.");
      }
    }
  }
```

Reply:
Yes, but it is done in the second PR.
It involved another set of changes that would make the PR too big (it's already not a small one ;) )

Reply:
Looks good in #996 👍